### ISSUE TITLE
Implement English-to-Cree search, based on WordNet senses

### ISSUE THREAD

**Issue by aarppe** [UAlbertaALTLab/morphodict#1138]
Here's a mock-up HTML version representing what an implementation of searching with English words would look like, when matched with WordNet senses under which the Cree entries would be organized based on their manual classifications according to WN (in some order, here in decreasing corpus frequency).

<html>
   <title>English WN sense &#8594; Cree CW entries: WAIT</title>
   <body>
      <ol>
         <li><b><a href="http://wordnetweb.princeton.edu/perl/webwn?s=wait">WAIT</a></b> (V): <em>stay in one place and anticipate or expect something</em>
            <br>[ (v) wait#1 (100) ]</li>
            <ul>
               <li><b><a href="https://itwewina.altlab.app/word/pêhêw/">pêhêw</a></b> (VTA-1)</li>
                  <ol>
                     <li>s/he waits for s.o.</li>
                  </ol>
               <li><b><a href="https://itwewina.altlab.app/word/pêhow/">pêhow</a></b> (VAI-1)</li>
                  <ol>
                     <li>s/he waits</li>
                     <li>s/he expects, s/he is in expectation (of someone's arrival)</li>
                  </ol>
               <li><b><a href="https://itwewina.altlab.app/word/pêhtâw/">pêhtâw</a></b> (VTI-2)</li>
                  <ol>
                     <li>s/he waits for s.t.</li>
                  </ol>
               <li><b><a href="https://itwewina.altlab.app/word/aswaham/">aswaham</a></b> (VTI-1)</li>
                  <ol>
                     <li>s/he waits eagerly for s.t., s/he lies in wait for s.t.</li>
                     <li>s/he catches s.t. as it drips</li>
                  </ol>
               <li><b><a href="https://itwewina.altlab.app/word/pa-pêhêw/">pa-pêhêw</a></b> (VTA-1)</li>
                  <ol>
                     <li>s/he is still waiting for s.o.</li>
                  </ol>
            </ul>
            <br>
         <li><b><a href="http://wordnetweb.princeton.edu/perl/webwn?s=wait">WAIT</a></b> (V): <em>look forward to the probable occurrence of</em>
            <br>[ (v) wait#3 (33) - (v) await#1 (25) - (v) expect#3 (12) - (v) look#8 (1) ]</li>
            <ul>
               <li><b><a href="https://itwewina.altlab.app/word/pêhow/">pêhow</a></b> (VAI-1)</li>
                  <ol>
                     <li>s/he waits</li>
                     <li>s/he expects, s/he is in expectation (of someone's arrival)</li>
                  </ol>
               <li><b><a href="https://itwewina.altlab.app/word/âswamêw/">âswamêw</a></b> (VTA-1)</li>
                  <ol>
                     <li>s/he awaits s.o. (as food) with an open mouth</li>
                  </ol>
               <li><b><a href="https://itwewina.altlab.app/word/âswahtam/">âswahtam</a></b> (VTI-1)</li>
                  <ol>
                     <li>s/he awaits s.t. (as food) with an open mouth</li>
                  </ol>
               <li><b><a href="https://itwewina.altlab.app/word/âswahcikêw/">âswahcikêw</a></b> (VAI-1)</li>
                  <ol>
                     <li>s/he awaits food with an open mouth (e.g. baby, bird)</li>
                  </ol>
            </ul>
            <br>
         <li><b><a href="http://wordnetweb.princeton.edu/perl/webwn?s=wait">WAIT</a></b> (V): <em>wait before acting</em>
            <br>[ (v) wait#2 (49) - (v) hold_off#2 (0) - (v) hold_back#4 (0) ]</li>
            <ul>
            </ul>
            <br>
         <li><b><a href="http://wordnetweb.princeton.edu/perl/webwn?s=wait">WAIT</a></b> (N): <em>the act of waiting (remaining inactive in one place while expecting something)</em>
            <br>[ (n) waiting#1 (2) - (n) wait#2 (2) ]</li>
            <ul>
               <li><b><a href="https://itwewina.altlab.app/word/pêhowin/">pêhowin</a></b> (NI-1)</li>
                  <ol>
                     <li>waiting</li>
                     <li>expecting someone</li>
                  </ol>
            </ul>
            <br>
         <li><b><a href="http://wordnetweb.princeton.edu/perl/webwn?s=wait">WAIT</a></b> (N): <em>time during which some action is awaited</em>
            <br>[ (n) wait#1 (2) - (n) delay#1 (2) - (n) time_lag#1 (0) - (n) postponement#1 (0) - (n) hold#4 (0) ]</li>
            <ul>
            </ul>
            <br>
         <li><b><a href="http://wordnetweb.princeton.edu/perl/webwn?s=wait">WAIT</a></b> (V): <em>serve as a waiter or waitress in a restaurant</em>
            <br>[ (v) wait#4 (2) - (v) waitress#1 (0) ]</li>
            <ul>
            </ul>
            <br>
      </ol>
   </body>
</html>
-----

The above is created with the following incantation, run in the ALTLab repo in: `altlab/eng/bin/`:

`./sort-cw-entries-by-wn-senses.sh ../../crk/dicts/Wolvengrey_WN_RW.tsv ../dicts/WordNet-3.0 ../../crk/generated/ahenakew_wolfart_bloomfield.fst+cg.freq-sorted.txt | ./organize-cw-wn-senses-by-wn-synsets.sh | ./pretty-print-wn-sense-search.sh 'wait' '' html | less
`

What this would require are the following:
* Most if not all Cree entries in all the dictionaries have been classified (manually or otherwise) according to WN senses
* All Cree entries with WN senses that belong to the same synset are grouped together (so we would need to incorporate the WN ontology).
* WN synsets are shown in their descending frequency (as known from the SEMCOR corpus, and included in the WN source data).
* Cree entries under synsets are sorted in some way. The above implementation uses descending corpus-frequencies from the AW+B corpus, but one might want to manually improve this.
* Potentially some Cree entries might be missing from some WN synset / sense. This can be fixed by adding the WN semantic class to the Cree entry in question.

**Comment by aarppe** [comment]
@nienna73 @DBDacanay I couldn't resist creating a full-fledged mock-up example case, illustrating what the English-to-Cree search could actually look like, describing the broad strokes and open questions in this meta-issue. When you need a break, you might consider whether the above would make sense, or might benefit from some further tweaking.

**Comment by fbanados** [comment]
I think it would be better to only show the wordnet senses that have words in the dictionary (in the example, omitting 3., 5., and 6.)

**Comment by fbanados** [comment]
Current example, sorts WN senses by frequency in WN corpus, and then by frequency in `lemma_freq.txt.  I've added a link to the next WN meaning just as there might be many words and definitions on each. We can discuss this when we meet. ![Screenshot 2024-12-03 at 11 41 53 AM](https://github.com/user-attachments/assets/79f74697-b3fd-47e5-9e47-dd091e766771)
![Screenshot 2024-12-03 at 11 43 04 AM](https://github.com/user-attachments/assets/7bc81b94-7188-40e1-995b-f6b035cf55df)
![Screenshot 2024-12-03 at 11 43 20 AM](https://github.com/user-attachments/assets/3a653314-843a-4eac-b89e-05ea24f9c7c7)

**Comment by aarppe** [comment]
There a version of a corpus-based reverse frequency file that I was intending to be used for morphodict purposes, having all sorts of information useful for other purposes as well, which has been created for the Ahenakew-Wolfart and Bloomfield corpora, but not yet to the full extent for the Supplementary corpus, cf.

```
cat generated/ahenakew_wolfart_bloomfield.fst+cg.freq-sorted.txt | cut -f3,4 | sort | uniq | sort -k1,1nr | gawk 'match($2, "[[:punct:]]")==0 && $2!=""' | head -15
6589	êkwa
3457	mâna
3214	awa
2669	êkosi
2595	itêw
2558	itwêw
2524	ôma
1751	mîna
1566	êkota
1300	ôhi
1207	êwako
1148	mistahi
1136	kîkway
1060	êsa
1057	mâka
```


It is this (type of) file that I'd use to get the lemma-frequencies. I've only picked the columns for the lemmas and their frequencies, though one might need the linguistic analysis field as well to get the POS-tags for disambiguation. Also, as discussed there would be the dictionary-based morpheme-aggregate frequencies, that we may also want to consider, as well as the glossary-frequencies, when ranking the Cree entries within each English-based synset.

**Comment by aarppe** [comment]
Other features to be explored:

1. Numbering the senses incrementally, per each part-of-speech, i.e. `1. wait (v)`, `1. wait (n)`, `2. wait (v), ...` 
2. Encoding the English parts-of-speech with relabeling, i.e. Code: `(n)` --> Linguistic: `Noun`, Plain English: `Naming word`, Plain nêhiyawêwin: `wihowin-itwêwin`, and so on. These should be added to the relabeling file.

**Comment by fbanados** [comment]
Added a draft encoding for relabellings (if changing before deployment to .app, please make sure changes are made in the dev branch and not in main)

https://github.com/UAlbertaALTLab/morphodict/blob/dev/src/morphodict/resources/crk.altlabel.tsv

**Comment by aarppe** [comment]
I can see the expand/collapse/forward/backward buttons, but they are not changing anything visually, as far as I can see.

**Comment by fbanados** [comment]
I've been debugging that situation and should be solved but I am in the process of deploying it

**Comment by fbanados** [comment]
> * [x]  Implement an option to use primarily symbols/icons to toggle features and navigate results. E.g. Use a down-arrow ⬇️ to hop to the next sense in WN search, plusses and minuses to expand/collapse accordion panes ➕/➖, etc.

Update with icons should be working now.  Do we want to have a text option or just the icons will suffice?

**Comment by aarppe** [comment]
Personally, I've been a fan of icons, as they could keep the output concise and minimalistic (in my mind), but some have liked textual explanations.

Perhaps we could have icons vs. text (not just for this, but generally) as an option (and/or seeing textual instructions when hovering over an icon)?

**Comment by fbanados** [comment]
> Perhaps we could have icons vs. text (not just for this, but generally) as an option (and/or seeing textual instructions when hovering over an icon)?

That could be done, maybe on a separate issue.  I'll start that.

**Comment by aarppe** [comment]
Of the remaining subtasks, getting the English phrase analysis/generation incorporated in the WN search would seem the one remaining worthwhile item - while I’ll have to regenerate the corpus frequency data, and we’ll need to add in the new CW entries as well as their WN classifications.

Among other longstanding issues, being able to link entries/wordforms/lemmas to Korp (opening a new tab. I’d think) would be another longstanding request. (#156)

**Comment by aarppe** [comment]
Taking a look at previous issues, the metaissue #516 seems like a good overview of possible next steps, and which of these we might prioritize (now check-marked).

**Comment by fbanados** [comment]
We should not wait on the ESPT FSTs to be solved before deploying WordNet senses and search into itwewina.
